### ISSUE TITLE
NetMQ + Unity delayed data

### ISSUE THREAD

**Issue by kornellvarga** [zeromq/netmq#682]
### Environment

    NetMQ Version:  Latest 3.x
    Operating System: WIn10
    .NET Version: 3.5

I have a working version of a C# code in Unity where the communication works perfectly between our server (Python based, I'm not part of the development) and the Client in Unity.
This works using a secondary thread in Unity, however for certain reasons we would like to switch to a Single threaded version.

### Expected behaviour
I expect to receive NetMQMessages with 2 frames. First frame is a type, second frame is a JSON datastructure. Realtime, with little to no delay from server to client.
For example a float value changes in the JSON gradually from 0.0 to 1.0 and I should see this data in Unity without delay, just as in the multithreaded version.

### Actual behaviour
I am receiving the two frames and the JSON datastructure, however with a great delay, and I believe I am also missing some messages. 
I am using a subscribersocket to ReceiveMultiPartMessage() on every loop (I have tried simply on Update, but also in a coroutine with an infinite loop where Im giving back the control to Unity after each dataprocess with yield return null).

What baffles me is repeating the ReceiveMultiPartMessage within the same loop does decrease the delay in receiving the data.

I might misunderstand some basic NetMQ behaviour, could anyone please enlighten me of what mistake of mine could this behaviour be a symptom?

Some further detailes:
I have browsed through the issues here, I hope I have not missed the solution to this problem.
I am using the mentioned Contect handling in 3.x, the AsyncIO.ForceDotNet.Force(); and I am also terminating the sockets, thread, and context

Thank you in advance for any assistance!

**Comment by somdoron** [comment]
I think the time between each iteration of the loop is long and this is why you get the delay.
I would suggest using a secondary thread and enqueuing the messages to the main thread through something like ISynchronizeInvoke.

**Comment by kornellvarga** [comment]
Thank you for the suggestion, however as I mentioned, I already have a multi-threaded solution that does work fine!
The reason why we are reworking it for a single threaded version is because this app is meant to run on UWP & Hololens which doesnt like Unity's multithreading

**Comment by somdoron** [comment]
NetMQ is multithread internally, so you don't really solve it.

If that is the case I would suggest calling TryReceive every X milliseconds, the delay should be up to X.

**Comment by kornellvarga** [comment]
Yes, I have read that NetMQ is internally multithread, however I had not chance yet to test a UWP build to see. I can imagine that only Unity's Thread handling causes the build problems.

I have also tried the TryReceiveMultiPartMessage(timeout, out message), with different timeout setups (on secondary threaded version successfully on single thread not) .

I'm not entirely sure if I understand the architecture of this method completely.
I used it on every loop, but your comment makes me think that I would have to run this from Unity only once and it automatically repeats on NetMQ's internal threading every X milliseconds?

Or you mean, solving it from a Unity loop, on the mainThread to call the TryReceiveMultiPartMessage(), multiple times, as long as the current loop takes longer than the X millisecond.

**Comment by Mystfit** [comment]
I had similar problems in the past when trying to make Unity play nice with
NetMQ and threads, since Unity doesn't like it when you try to manipulate
it's objects from outside of the main Unity thread.

My solution was to add a delegate to the incoming socket that took the
incoming message off of the socket and placed it in a ConcurrentQueue. Then
in a MonoBehaviour Update function I would dequeue each item with a while
loop until the queue was empty. This let the main Unity thread perform any
actions using the messages that it wished, without complaining about
illegal thread access.

You could also use a coroutine that would yield until there was at least 1
item in the queue, pop every item off until the queue was empty and then do
something interesting.

In regards to whether or not NetMQ itself would function inside a Hololens
and UWP I have no idea. I don't see why not though, since it seems that UWP
does support System.Threading as well as thread pooling

**Comment by kornellvarga** [comment]
@Mystfit  Have you ever successfully built a project switching to Windows Store platform in Unity using Threading?
I am constantly facing some errors on building the project, I have just reproduced it with a clean project.
I have tried both delegating methods to the ThreadStarand simply using a method and running it as a new thread like : _thread = new Thread( NetMQClient );

About handling the threads in Unity I do not have any problem, the project works 100% perfect on my multithreaded architecture, which I am trying to turn into a single threaded version to tackle the UWP issues. I am using a Task queuing to communicate between the threads.

Could you please elaborate a bit on the delegate solution you made, im curious how that looks like. I mean I know how to build a queue, but what exactly do you mean by delegating to the incoming socket?

**Comment by kornellvarga** [comment]
@somdoron  What do you think about this solution?

**Comment by Mystfit** [comment]
@kornellvarga Sorry, I should have clarified further. The delegate was just
a function that would get called when a socket received data when attached
to a NetMQPoller. I'd attach the function to the ReceiveReady attribute on
the socket and this would be called by the poller whenever the socket
received data. It was still a threaded solution, since the poller would
block until it received a message.
Also, I haven't tried to build any Unity projects with Windows store
support sorry. What version of Unity are you running? If possible could you
give the Unity2017 beta a go?

**Comment by kornellvarga** [comment]
@Mystfit  Yes, I was thinking the same that I will try Unity 2017, especially because with that version finally I could switch to NetMQ 4.x .
I am not really familiar with the NetMQPoller, but I am going to look into some documentation, seems like a good solution. 
Would this mean that the NetMQPoller acts as a secondary thread, is this the "internal" threading @somdoron  mentioned? 
If so, and I do not have to use Unity's built-in threading, the NetMQ's poller thread might be handeled differently by the Windows Store build configuration.

Is [this ]( http://netmq.readthedocs.io/en/latest/ ) the best source for learning about NetMQPoller usage?

Thx a lot!!

**Comment by kornellvarga** [comment]
@Mystfit  I think I have figured it out. So, in our case there is a subport that I have to grab first, that I do with an initial conneciton, later I handle subscription with a SubscriberSocket and I delegate to ReceiveReady on the subscribersocket.
` 
void OnConnected(){
	NetMQMessage mgs = new NetMQMessage();

	subscriberSocket.ReceiveReady += (s,a) => {

		int ind = 0;

		while(a.Socket.TryReceiveMultipartMessage(ref msg) && ind <= 10){
			ind++;
			GatherData(msg);
		}


	}
}
void Update(){
	subscriberSocket.Poll();
}
 `

What do you think? So far it seems pretty responsive, and on one Unity thread.

**Comment by JeffBail** [comment]
@kornellvarga did you end up trying Unity 2017 and is it working better for you than 5.x?

**Comment by jwvanderbeck** [comment]
All of this threading has me thoroughly confused.  This is an area of programming that I have little to no experience with.

Can someone bottom line it for me? Is it at all possible to wrk with NetMQ without threading in Unity? It NetMQ itself is threaded that doesn't affect me.

I am tasked with developing some Unity Editor tools that need to talk with an existing ZMQ server and I am pulling my hair out trying to figure it out.

I had something that vaguely works, but for some reason if I make a NetMQ call, every time Unity reloads assemblies after that, such as after code changes, the editor Freezes.

**Comment by oocpawel** [comment]
If Your Unity Edtor is freezing after play, when You are using NetMq, it looks like sockets wasnt cleared properly, try to call NetMQConfig.Cleanup(false); after You call Dispose on socket

**Comment by somdoron** [comment]
I'm afraid I don't have much Unity experience, I only played with it few times.
Anyway, NetMQ is using threads, NetMQPoller can be run on a different thread (RunAsync) or the current thread and block it (Run). All of the threads run on the user space, so you are using Unity with threads.

**Comment by cz2011301070** [comment]
> Thank you for the suggestion, however as I mentioned, I already have a multi-threaded solution that does work fine! The reason why we are reworking it for a single threaded version is because this app is meant to run on UWP & Hololens which doesnt like Unity's multithreading

Did u make it run on hololens with Netmq?
